### ISSUE TITLE
let do_excel_import() update existing records instead of failing with an uninformative message

### ISSUE THREAD

**Issue by i92guboj** [opensourcepos/opensourcepos#1031]
This discussion started in #1026 

Hi!

I'd like to be able to update elements using csv import, and while I am at it, simplify a bit the logic behind this function and related ones.

Right now, as in ospos 3.0.1, import works in this way:

A) do_excel_import() iterates over the csv array
B) it assigns values, does stuff... then it checks if data[0] (item_number wannabe) is empty which marks it as "do not import this one" (sigh, this could be done the first thing of all I guess and we'd simplify logic quite a bit), but have to revise the code slowly and double check.
C) if the item is not empty (!invalidated) then try to insert it, little error checking and no feedback at all to the end user here which is quite a lot of fun to blindly debug when you are trying to import something and scales pretty badly when the listing of products grows

I propose:
A) iterate the csv array
B) if item is empty, directly skip it, nothing else
C) modify the logic, lookup the list, if item exists update it, else import it as always, this might need deeper changes, only looked at the code superficially...
D) this should produce less "unimportable" rows
E) do something about the blind-ness, users can't fix errors they know nothing about, can they?

Key for this logic will be that item_number is unique, and can't be empty. So:

@jekkos , in #1026  you say:

> There might be some usecases for items without item_numbers, which are eg service items, like hours of labor or generic items that represent products not counted in stock anymore. Currently all items have stock counter linked and that might change in the future as described in #200

I fail to see the picture. Could you, please, develop a bit more on this? Why do they need an empty item_number? I just checked, and that field is unique at sql level, at least in my installation. You can't certainly have two rows with item_number='<empty>' currently.

I also don't get why serviceable items need that, surely I am missing something obvious, but... why or how to item_numbers relate at all with stock? You can surely have all items tied to an item_number, with independence on whether they are material or not. You need that if you plan to use a barcode reader as well, which incidentally I do. 

Import ALREADY REQUIRES item_number to be non-empty, and unique.

Ospos UI to create items ALREADY REQUIRES  item_number to be non-empty, and unique, and even warns you if it isn't unique.

I'll read #200 slowly, though. I am surely missing something trivial here.

However, PKs are nice, but without a stable and unique ID that can be used to connect the db to the external real world, we will never be able to correlate items inside ospos with those inside the shelves, in my shop. ;)

Last and least, updating thousands of products by hand each time I make an offer or discount is not an option. I really wonder how people live without this feature. Don't you have black friday-esque events around there? Don't you offer discount when changing season? How does the average Joe works around that without having to apply maths into tables using SQL? You can't certainly mass-edit prices, unless you own some kind of "everything for 10 bucks" kind-of shop.

Just bear with me, I want to iron a bit this one and see if we can find some middle ground to build on this without screwing up anything else in ospos :)

**Comment by i92guboj** [comment]
Also, @jekkos 

> OK just a short answer for the following
> 
> B) not sure whether it's mandatory in the import. I know it's certainly not in the database tables, as items can also be scanned using their Mysql PK. If the code does not enforce an item_number right now then it could break some functionality for other users out there. Code needs to be reviewed and depending on the outcome we'll need to decide what to do with it

item_number is unique, and, hance, mandatory. It just happens that ONE (and only ONE) of the rows, can have the contents '<empty>'. It won't let you insert a second one with the same contents, though, since that'll violate unique-ness.

> C) Solution here depends on the former, do we require item_number on import or not? If yes then indeed we can always decide insert/update..

It does require an item_number already. So we are already at that juncture. We only need to make it worth. If you update an existing element all that will happen is that you'll get one less error in your output ;)

Of course, it could happen that users will accidentally overwrite rows. That's fine for me. When I mass dump a table into a database I take for granted that's my responsibility to make sure the input is correct, and it's also my responsibility to backup.

However, I think we could add a checkbox to [x]skip or [x]overwrite existing items in the import dialog. That'll be a good thing, I guess. Taking this into account, it would also be nice to get a report about what's been done during the process. And it shouldn't be too hard for what's I've seen while tinkering about in that part of the code.

**Comment by jekkos** [comment]
@i92guboj I think we missed the case of item_number being NULL, which as far as I can see might yield a successful import as $invalidated might remain FALSE. I haven't tested in practice and am not certain if getcsv() will produce an empty string or null in case it encounters an empty cell? Think we should double check that first before we can confirm that items without item_number can be imported?

I do have a usecase in which I have items without item_number. Barcodes can be generated for these kind of items. In absence of item_number it will fallback to the item_id as seed. So just one more quick check and then I think we can go ahead with the implementation?

Thanks!

**Comment by jekkos** [comment]
Good idea by the way on the overwrite and skip part. Making the process interactive will surely benefit the user experience here. Would be nice to have that as an improvement for the project!

**Comment by i92guboj** [comment]
Just a quick note. I am fairly sure that it can't import two empty item_numbers. So, it translates empty cells to '', not NULL. I can't check right now. 

I have to analyze the rest of the info you just gave me. This can potentially break lots of things. But, I'd like to finish #1026 before getting into this.

**Comment by jekkos** [comment]
Ok could be right, I don't know by heart. If this is the case then your proposal is fine and I think we'd be good to proceed implementing the update logic.

**Comment by jekkos** [comment]
I have temporarily set the milestone to 3.1.0. Let's see how this goes.

**Comment by daN4cat** [comment]
I just skimmed through this but a rework of the excel import with some smarter way of re-conciliating items would be a good addition.

**Comment by i92guboj** [comment]
In this regard, I have some ideas which would need discussion if I ever give this a try, but it's not something that will get priority for me. I have some other things to do before that, and this is already usable enough for my use case. I could benefit from a better import machanism, sure! but first I need to get ospos synchronising to drupal/ubercart, for example.

**Comment by jekkos** [comment]
For the sync story maybe have a look at our idea to add a REST API in #615. Would be nice to see some contribution to the project in that scope. Thanks!

**Comment by i92guboj** [comment]
Yes, I saw that and have bookmarked it. That'd be the way to go, I guess. Thanks!

**Comment by daN4cat** [comment]
I have a sync with drupal commerce based on SQL feed import. Basically Drupal commerce reads every 3h the status from OSPOS  database and syncs. I'm sure there could be better options.

**Comment by i92guboj** [comment]
At the end of the day it's just a matter of geting numbers from table A into table B. Solutions implemented by programmers usually lack fancy UI's and tend to be scripted rather than compiled, because, well, the text editor is your GUI.

The tedious part of it is mapping the tables and keeping the sync up with updates on both sides. But that same problem will always exist, no matter what paradigm you use to synchronize.

**Comment by daN4cat** [comment]
@i92guboj yes, the sync issue between the two table stays. In my wife's case they basically made OSPOS transactions as soon as possible after having received Drupal Commerce sell notification. So the master for the stock was always OSPOS. Drupal Commerce is just a replica and with a sync every 3h there is plenty of time to manage things properly.
But if you have many online transactions then it's better to make Drupal Commerce be POS itself and create the proper sale transaction flow in OSPOS. In that case RESTful API is required. We would need the same for a Mobile app as @jlctmaster mentioned in any discussion thread.

**Comment by daN4cat** [comment]
No idea what we want to do with this, but I saw no progress.

**Comment by i92guboj** [comment]
I have a clear idea on what _I_'d need, which mostly involves a way to mass update arbitrary fields for existing records, something every other POS does. It could be an "insert/update" selector, or whatever else. However don't hold up your breath for me doing this, because I am quite busy with other business issues. We are in the middle of a transition, and have a million other things to worry about. I handle mass modifications and complex insertions using SQL which I am proficient with, and so this is quite secondary. I guess you all do that the same way and that's why this functionality doesn't get much attention.

So, if you need to close this issue, please, do it, since I have no idea if/when I will have the time to put my hands on this. :(

**Comment by daN4cat** [comment]
Let's make #1412 the one that will implement this feature if @SteveIreland feels like going for it in 3.3.0

**Comment by objecttothis** [comment]
Duplicate of #1412
